source/extensions/transport_sockets/tls/downstream_context_config_factory.cc

```diff
@@ -0,0 +1,22 @@
+#include "extensions/transport_sockets/tls/downstream_context_config_factory.h"
+
+#include "extensions/transport_sockets/tls/context_config_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+Ssl::ContextConfigPtr DownstreamContextConfigFactory::createSslContextConfig(
+    const Protobuf::Message& config,
+    Server::Configuration::TransportSocketFactoryContext& context) {
+  return std::make_unique<ServerContextConfigImpl>(
+      dynamic_cast<const envoy::api::v2::auth::DownstreamTlsContext&>(config), context);
```

Comment:
If we universally do a dynamic_cast here why not just pass this type as a parameter?

Reply:
createSslContextConfig() is public interface which will also be used on client side which should downcast message to UpstreamTlsContext.

Reply:
@lizan any thoughts on this one? I would need to sit down and page in all of this code but something seems off here. Maybe this is related to your other comment.

Reply:
at interface level this seems fine, though at higher level I think this should just be a `TransportSocketConfigFactory`